### PR TITLE
Rds/kjw/function get building segment swh bat

### PR DESCRIPTION
**Pull request by KarenWGard**

### CODE REVIEW

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+- **building_segment**
+
+Returns:
+- **building_segment_swh_bat**: one of the ServiceWaterHeatingSpaceOptions2019ASHRAE901 options
+
+Function Call:
+
+- get_energy_required_to_heat_swh_use  
+- get_SWH_uses_associated_with_each_building_segment  
+- get_SWH_uses_associated_with_each_building_segment
```
Comment (claperle) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:16]: Repeated function

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            - get the other key.  First get a list of all keys: `all_keys = list(swh_use_dict.keys())`
+            - remove "UNDETERMINED": `all_keys.remove("UNDETERMINED")`
+            - set other_key to the remaining item in the list: `other_key = all_keys[0]`
+            - now check whether the water use for UNDETERMINED is < the water use for the other key: `if swh_use_dict["UNDETERMINED"] < swh_use_dict[other_key]:`
+                - set the building_segment_swh_bat equal to other_key: `building_segment_swh_bat = other_key`
+
+- return result: `return: building_segment_swh_bat`
+
+
+**Returns** swh_and_spaces_dict
```
Comment (claperle) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:53]: This needs to be updated or deleted

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    - create a dictionary that will hold the different types of swh_use_bat_types and the total service water used for the year: `swh_use_dict = {}`
+    - get the service water heating uses in the building segment `service_water_heating_use_ids = get_SWH_uses_associated_with_each_building_segment(RMD, building_segment.id)`
+    - look at each service water heating use id: `for swh_use_id in service_water_heating_use_ids:`
+        - get the swh_use using get_component_by_ID: `swh_use = get_component_by_ID(RMD, swh_use_id)`
+        - if any swh_use has use_units equal to "OTHER", the total energy required to heat the use cannot be determined, and we return "UNDETERMINED": `if swh_use.use_units == "OTHER": return "UNDETERMINED"`
+        - calculate the total hot water used using the function get_energy_required_to_heat_swh_use and add it to the swh_use_dict: `swh_use_dict[swh_use.area_type] += get_energy_required_to_heat_swh_use(swh_use, RMD)`
+        - check to see if the swh_use has service_water_heating_area_type: `if swh_use.area_type:`
+            - add the SWH building area type to the swh_use_dict and set the value to 0: `swh_use_dict.set_default(swh_use.area_type, 0)`
```
Comment (claperle) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:34]: If none of the swh_use BAT types are defined for a space, do we want to check to see if the space.service_water_heating_area_type was defined and use that in that specific case?

Comment (KarenWGard) [reply docs/ruleset_functions/get_building_segment_SWH_bat.md:34]: I was not checking **space.service_water_heating_area_type** because that doesn't make a lot of sense to me if we are moving towards assigning service water heating uses at the building segment level instead of at the space level, though it certainly could be done.  Thoughts?

Comment (claperle) [reply docs/ruleset_functions/get_building_segment_SWH_bat.md:33]: If it wouldn't be much additional effort, just to have anther way to establish the BAT for the segment, I think it could be a good third option.

Comment (KarenWGard) [reply docs/ruleset_functions/get_building_segment_SWH_bat.md]: ok, I did this in the simplest way possible while still maintaining encapsulation - I made a change to `get_energy_required_to_heat_swh_use`, so that it now returns a dictionary with space_ids as keys and energy used for each space as values.  For most uses, getting the total energy can be done by simply summing the values.  But for this function it allows us to look at the space BAT if needed.

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+- **building_segment**
+
+Returns:
+- **building_segment_swh_bat**: one of the ServiceWaterHeatingSpaceOptions2019ASHRAE901 options
+
+Function Call:
+
+- get_energy_required_to_heat_swh_use  
+- get_SWH_uses_associated_with_each_building_segment    
+- get_component_by_ID
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:16]: is this different from get_obj_by_id?

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+- get_component_by_ID  
+
+Data Lookup: None
+
+Logic:
+
+- set the building_segment_swh_bat to nil: `building_segment_swh_bat = "UNDETERMINED"`
+- if the building_segment.service_water_heating_building_area_type exists set the swh_bat to the one given: `if building_segment.service_water_heating_building_area_type:`
+    - set the building_segment_swh_bat to the building_segment.service_water_heating_building_area_type: `building_segment_swh_bat = building_segment.service_water_heating_building_area_type`
+- otherwise, if the building segment doesnt have service_water_heating_building_area_type, we need to determine the building segment SWH type by looking at individual SWH uses: `else:`
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:25]: It will be very helpful to avoid using terms like "use type", and "SWH type". They are confusing terms because there is `ServiceWaterHeatingUse.water_serves_type` and `ServiceWaterHeatingUse.area_type`. Please be consistent in referring to it as "area type"

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        - otherwise: `else:`
+            - go through each space served by the swh_use and see if it has a service_water_heating_bat: `for space_id in swh_use_energy_by_space:`
+                - get the space: `space = get_component_by_ID(RMD, space_id)`
+                - check if the space has a swh_use_bat: `if space.service_water_heating_bat:`
+                    - First add the BAT and set the default: `swh_use_dict.set_default("space.service_water_heating_bat", 0)`
+                    - add the energy used to the dict: `swh_use_dict["space.service_water_heating_bat"] += swh_use_energy_by_space[space_id]`
+                - otherwise, we'll add this use to UNDETERMINED.`else:`
+                    - First add UNDETERMINED and set the default: `swh_use_dict.set_default("UNDETERMINED", 0)`
+                    - add the energy used to the dict: `swh_use_dict["UNDETERMINED"] += swh_use_energy_by_space[space_id]`
+    - now we need to determine the building_segment swh_use_type based on the following rules:
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:44]: area type

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    -     1. At least 50% of the SWH uses needs to be assigned a use type
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:45]: area type

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    -     2. All SWH needs to be assigned to the same use type
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:46]: area type

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    - if there is only one element in swh_use_dict, the building_segment_swh_bat is the only key in swh_use_dict: `if(len(swh_use_dict)==1): building_segment_swh_bat = list(swh_use_dict.keys())[0]`
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:47]: if there is only one element in swh_use_dict, all spaces have the same service water heating area type

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    - otherwise, if swh_use_dict has two elements: `elsif(len(swh_use_dict)) == 2:`
+        - in order to meet the requirement that all SWH needs to be assigned to the same use type - ie any SWH that doesn't have a use-type is covered by "UNDETERMINED", we expect the dict to have keys "UNDETERMINED" and one other area use type: `if "UNDETERMINED" in swh_use_dict:`
```
Comment (JacksonJ-KC) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:49]: I found this confusing. Suggest rewording:

A building segment's service water heating area type cannot be determined if there is more than one key in the swh_use_dict (other than the "UNDETERMINED" key). If the "UNDETERMINED" key exists, it cannot be >= 50% of the use.

**Review by JacksonJ-KC — APPROVED** [review]

File: docs/ruleset_functions/get_building_segment_SWH_bat.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+- get_energy_required_to_heat_swh_use  
+- get_SWH_uses_associated_with_each_building_segment    
+- get_component_by_ID  
+
+Data Lookup: None
+
+Logic:
+
+- set the building_segment_swh_bat to nil: `building_segment_swh_bat = "UNDETERMINED"`
+- if the building_segment.service_water_heating_building_area_type exists set the swh_bat to the one given: `if building_segment.service_water_heating_building_area_type:`
```
Comment (yunjoonjung-PNNL) [inline docs/ruleset_functions/get_building_segment_SWH_bat.md:23]: I can't find the ```service_water_heating_building_area_type``` under ```building_segment```. Can you clarify this?

Comment (JacksonJ-KC) [reply docs/ruleset_functions/get_building_segment_SWH_bat.md:23]: https://github.com/open229/ruleset-model-description-schema/issues/264

**Review by claperle — APPROVED** [review]
